centreon/src/Core/HostCategory/Infrastructure/Repository/DbWriteHostCategoryActionLogRepository.php

```diff
@@ -0,0 +1,284 @@
+     *
+     * @param HostCategory $initialHostCategory
+     * @param HostCategory $updatedHostCategory
+     *
+     * @return array{
+     *    hc_name?: string,
+     *    hc_alias?: string,
+     *    hc_activate?: string,
+     *    hc_comment?: string
+     * }
+     */
+    private function getHostCategoryDiff(
+        HostCategory $initialHostCategory,
+        HostCategory $updatedHostCategory
+    ): array {
+        $diff = [];
+        $reflection = new \ReflectionClass($initialHostCategory);
+        $properties = $reflection->getProperties();
+
+        foreach ($properties as $property) {
+            $property->setAccessible(true);
+            $value1 = $property->getValue($initialHostCategory);
+            $value2 = $property->getValue($updatedHostCategory);
+
+            if ($value1 !== $value2) {
+                if ($property->getName() === 'isActivated') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = $value2 ? '1' : '0';
+                    continue;
+                }
+                if ($property->getName() === 'comment') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = $value2 ?? '';
+                    continue;
+                }
+                if ((string) $property->getType() === 'string') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = is_string($value2)
+                        ? $value2
+                        : throw new RepositoryException('Property value is not a string');
+                    continue;
+                }
+            }
+        }
+
+        return $diff;
+    }
+
+    /**
+     * Format the Host Category as property => value array.
+     *
+     * @param NewHostCategory $hostCategory
+     *
+     * @return array<string, string>
+     */
+    private function getNewHostCategoryAsArray(
+        NewHostCategory $hostCategory,
+    ): array {
+        $reflection = new \ReflectionClass($hostCategory);
+        $properties = $reflection->getProperties();
+
+        $hostCategoryAsArray = [];
+        foreach ($properties as $property) {
+            $property->setAccessible(true);
+            if ($property->getName() === 'isActivated') {
+                $hostCategoryAsArray[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]]
+                    = $property->getValue($hostCategory) ? '1' : '0';
+            } elseif ($property->getName() === 'comment') {
+                $hostCategoryAsArray[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]]
+                    = $property->getValue($hostCategory) ?? '';
+            } else {
+                $hostCategoryAsArray[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]]
+                    = is_string($property->getValue($hostCategory))
+                        ? $property->getValue($hostCategory)
+                        : throw new RepositoryException('Property value is not a string');
```

Comment:
Please note that this is not a repository error, but a business rule error for action logs. The decorator should not have the same exception as the repository. It should be an ActionLogException, for example.

Reply:
Seen together, I resolve.